lib/src/front_end/ast_node_visitor.dart

```diff
@@ -476,66 +478,114 @@ class AstNodeVisitor extends ThrowingAstVisitor<void>
+        //
+        // ```
+        // for (
+        //   initializerClause;
+        //   conditionClause;
+        //   incrementClause
+        // ) {
+        //   body;
+        // }
+        // ```
+        var partsList =
+            DelimitedListBuilder(this, const ListStyle(commas: Commas.none));
+        partsList.leftBracket(node.leftParenthesis);
+
+        // The initializer clause.
+        if (initializer != null) {
+          partsList.addCommentsBefore(initializer.beginToken);
+          visit(initializer);
+        } else {
+          // No initializer, so look at the comments before `;`.
+          partsList.addCommentsBefore(forParts.leftSeparator);
+        }
+
+        token(forParts.leftSeparator);
+        partsList.add(pieces.split());
+
+        // The condition clause.
+        if (forParts.condition case var conditionExpression?) {
+          partsList.addCommentsBefore(conditionExpression.beginToken);
+          visit(conditionExpression);
+        } else {
+          partsList.addCommentsBefore(forParts.rightSeparator);
+        }
 
-      case ForPartsWithDeclarations():
-        partsList.addCommentsBefore(forLoopParts.variables.beginToken);
-        visit(forLoopParts.variables);
-        finishForParts(forLoopParts, partsList);
+        token(forParts.rightSeparator);
+        partsList.add(pieces.split());
 
-      case ForPartsWithExpression(:var initialization?):
-        partsList.addCommentsBefore(initialization.beginToken);
-        visit(initialization);
-        finishForParts(forLoopParts, partsList);
+        // The update clauses.
+        if (forParts.updaters.isNotEmpty) {
+          partsList.addCommentsBefore(forParts.updaters.first.beginToken);
+          createList(forParts.updaters,
+              style: const ListStyle(commas: Commas.nonTrailing));
+          partsList.add(pieces.split());
+        }
 
-      case ForPartsWithExpression():
-        // No initializer part.
-        partsList.addCommentsBefore(forLoopParts.leftSeparator);
-        finishForParts(forLoopParts, partsList);
+        partsList.rightBracket(node.rightParenthesis);
+        pieces.split();
+        forPartsPiece = partsList.build();
 
       case ForPartsWithPattern():
-      case ForEachPartsWithDeclaration():
-      case ForEachPartsWithIdentifier():
+        throw UnimplementedError();
+
+      case ForEachParts forEachParts &&
+            ForEachPartsWithDeclaration(loopVariable: AstNode variable):
+      case ForEachParts forEachParts &&
+            ForEachPartsWithIdentifier(identifier: AstNode variable):
+        // If a for-in loop, treat the for parts like an assignment, so they
+        // split like:
+        //
+        // ```
+        // for (var variable in [
+        //   initializer,
+        // ]) {
+        //   body;
+        // }
+        // ```
+        token(node.leftParenthesis);
+        visit(variable);
+
+        finishAssignment(forEachParts.inKeyword, forEachParts.iterable,
```

Comment:
I'm not a fan of having `finishAssignment` or `AssignPiece` to describe this splitting type. It's kinda confusing to read and I wish we had a better way to describe it like `finishOperatorExpression` or `OperatorExpressionPiece`. Maybe not that, but it seems odd to read it as an assign or an "assignment" style.

Reply:
I'm going to go ahead and land this, but I'm definitely up for iterating on names.